### DISCUSSION TITLE
Add functionality to tilemap that automatically sets precise colisions in the shape of the tile.

### DISCUSSION THREAD

**Discussion by RakibRyan** [godotengine/godot-proposals#9630] [Editor]
### Describe the project you are working on

a tilemap based game where a lot of tiles aren't square.

### Describe the problem or limitation you are having in your project

When working with non square tiles manual collision shape setup becomes tedious. 

### Describe the feature / enhancement and how it helps to overcome the problem or limitation

add a button in tilemap that when pressed the physics layer in tile autosets to shape of the tile. 

### Describe how your proposal will work, with code, pseudo-code, mock-ups, and/or diagrams

it could be like unity's auto colission shape.

### If this enhancement will not be used often, can it be worked around with a few lines of script?

the enhancement will be used often if people had access to it. 

### Is there a reason why this should be core and not an add-on in the asset library?

it's too small to be it's own asset and too tedious to not be in core engine.

**Comment by AThousandShips** [comment]
Please explain some more details on what this would mean. Would it detect non-transparent pixels? Should it be strictly convex or concave as well? How would yo activate this? People aren't guaranteed to have used unity or this specific feature so please explain it instead

**Comment by RakibRyan** [comment]
it would trace the outlines of the image and create a colission shape based on it's outlines. kind of like this 
![maxresdefault](https://github.com/godotengine/godot-proposals/assets/85598032/cc52cfa3-4c47-4c3d-972d-8d68d896d1fe)

not expanding outwards but the exact shape the image is gonna be.  which will save a lot of time in adding precise collision to lot of assets where needed.


Note: i'm not just saying put it would have to create a thin collision line. I'm just saying whatever the shape we give it it would be able to create a precise colission based on it's silhouette.  if it's a triangle it would be able to create a triangle shaped colission if it's a decahedron or some other complex shape it would be able to automatically match to that shape if the user tells it to. i'm just proposing to bake that kind of functionality to engine's tilemap. and a button that's labeled "match sprite"/ something fitting like that. hope language barrier doesn't prevent it from me explaining it.

**Reply by AThousandShips** [reply]
So to clarify, it would be based on the outline of the *visible* parts, i.e. with a transparent background? It wouldn't, for example, work with a red shape on a white background

**Reply by RakibRyan** [reply]
yes i mean transparent background. do you import assets that are red with white background or that are transparent background?

**Reply by AThousandShips** [reply]
I didn't import anything :) I'm just asking again because it was one of my questions above that was intended to clarify what you were asking for since you gave so little detail

**Reply by RakibRyan** [reply]
my apologies. I hope I was able to make it clear now.

**Comment by aXu-AP** [comment]
It could be doable since there's already a similiar functionality for Sprite2D (accessible by selecting a Sprite2D and selecting from toolbar Sprite2D -> Create CollisionPolygon2D Sibling).
I think it would be placed in this menu here (something like "generate polygons from transparency"):
![kuva](https://github.com/godotengine/godot-proposals/assets/1621768/698d5493-bddf-4052-b6ae-b30477f87b97)
Altough the option should propably be compatible with tile selection as well, since the user might not want to generate collision for all tiles.

This would open a dialog with following options:
* Select physics layer
* Select navigation layer (creates inverted polygon)
* Select occlusion layer
* Same options as in sprite->polygon tool (simplification, shrink, grow)

However, I think there could be a potential problem in how to handle slopes. Take this for example:
![kuva](https://github.com/godotengine/godot-proposals/assets/1621768/5432f174-80b1-4853-92df-914c36cca512)
How the algorithm should determine where to put this point to prevent this step? For tile corners a special parameter ("snap to corners distance") could be provided, but for this kind of middle points its harder. One possibility is to ensure connectivity between adjacent tiles, but this would add a special requirement for how the tiles are arranged in the texture. Any ideas how to overcome this problem?

**Comment by RakibRyan** [comment]
I was thinking more like this. The shape itself would be the colission mask. I know it terms of performance it would be costly to do so precise colissions. But there are places it'd make sense. Ex: if you are trying to make the player have precise colission, or like if you want to make those complex shapes and are tedious and would need precise colissions.

![complexshort](https://github.com/godotengine/godot-proposals/assets/85598032/2f05c3d6-94d0-4b0e-9070-6a4997fa7929)

**Reply by aXu-AP** [reply]
I wonder what kind of tileset would have so exotic shapes? I mean, what I suggested would work even with such shapes. I just gave the most common example where this would be used, slopes, and what problems could arise.

**Reply by AThousandShips** [reply]
This would be very much dependent on what the user is wanting for these kinds of strange shapes, would they want a convex hull, or just a partial hull without too much detail, etc.

Though I don't think it's representative of almost any use case of this feature
